[JENKINS-37231] CORS allowed origins list never updated

escoem opened the pull request:
This is only a fix for JENKINS-37231 so I haven't refactor the plugin code. I will do later.

My goals have been reload list only on each configuration change, not always the list is requested.

List moved to DescriptorImpl as transient.
Added reload method (test requires)

escoem commented:
Please @christ66 @uaarkoti merge and release if you consider.

I will send a new PR soon with a full refactor of the plugin.

escoem commented:
Hello @christ66, any comments before merging?

escoem commented:
I won't send if you think is not required @christ66 

Anyway, can you consider this PR that fixes https://issues.jenkins-ci.org/browse/JENKINS-37231 please?

recampbell commented:
@reviewbybees

reviewbybees commented:
This pull request originates from a [CloudBees](https://www.cloudbees.com/) employee. At CloudBees, we require that all pull requests be reviewed by other CloudBees employees before we seek to have the change accepted. If you want to learn more about our process please see [this explanation](https://github.com/reviewbybees/about#about-reviewbybees).

recampbell commented:
@reviewbybees done

christ66 commented on src/main/java/org/jenkinsci/plugins/corsfilter/AccessControlsFilter.java:
```diff
@@ -34,7 +34,6 @@
     public static final DescriptorImpl DESCRIPTOR = new DescriptorImpl();
     private static final Logger LOGGER = Logger.getLogger(AccessControlsFilter.class.getCanonicalName());
     private static final String PREFLIGHT_REQUEST = "OPTIONS";
-    private List<String> allowedOriginsList = null;
```
:bug: I think this field should be marked as `transient` instead of being removed. This will most likely cause xstream issues when trying to load this object.

escoem replied:
This is not part of the configuration but a list created the first time the isAllowed method is executed. No xstream issue.

oleg-nenashev replied:
This object is not persisted, so it should be fine. The code is not ideal, but this removal should be safe

christ66 commented on src/main/java/org/jenkinsci/plugins/corsfilter/AccessControlsFilter.java:
```diff
@@ -147,9 +137,11 @@ public DescriptorImpl getDescriptor() {
         private String allowedHeaders;
         private String exposedHeaders;
         private String maxAge;
+        private transient List<String> allowedOriginsList;
```
:bug: I don't think this should be marked as `transient` since we need the contents of this variable saved to disk.

escoem replied:
We don't need to store that in the configuration file because of is already stored in private String allowedOrigins;

oleg-nenashev replied:
🐛 Without readResolve() logic this code is going to contain `null` after the configuration reload and hence it will fail with NPE in `isAllowed() `

escoem replied:
thanks @oleg-nenashev but IIUC, is not required because of it's populated inside the constructor:

              allowedOriginsList = createAllowedOriginsList(allowedOrigins);

Anyway, I'm going to add the readResolve method as per your request.

christ66 commented:
See this wiki on retaining backwards compatibility: https://wiki.jenkins.io/display/JENKINS/Hint+on+retaining+backward+compatibility#Hintonretainingbackwardcompatibility-Scenario:Removeafield

escoem commented:
Hi @christ66 

This is the config file `org.jenkinsci.plugins.corsfilter.AccessControlsFilter.xml` with current version (1.1)

```
<?xml version='1.0' encoding='UTF-8'?>
<org.jenkinsci.plugins.corsfilter.AccessControlsFilter_-DescriptorImpl plugin="cors-filter@1.1">
  <enabled>true</enabled>
  <allowedOrigins>http://foo.a, http://bar.b, http://acme.org</allowedOrigins>
  <allowedMethods>GET, PUT, OPTIONS</allowedMethods>
  <allowedHeaders>HEADER_1</allowedHeaders>
  <exposedHeaders>X-My-Custom-Header, X-Another-Custom-Header</exposedHeaders>
  <maxAge>120</maxAge>
</org.jenkinsci.plugins.corsfilter.AccessControlsFilter_-DescriptorImpl>
```

And this is the same file using my proposal:

```
<?xml version='1.0' encoding='UTF-8'?>
<org.jenkinsci.plugins.corsfilter.AccessControlsFilter_-DescriptorImpl plugin="cors-filter@1.2-SNAPSHOT">
  <enabled>true</enabled>
  <allowedOrigins>http://foo.a, http://bar.b, http://acme.org</allowedOrigins>
  <allowedMethods>GET, PUT, OPTIONS</allowedMethods>
  <allowedHeaders>HEADER_1</allowedHeaders>
  <exposedHeaders>X-My-Custom-Header, X-Another-Custom-Header</exposedHeaders>
  <maxAge>120</maxAge>
</org.jenkinsci.plugins.corsfilter.AccessControlsFilter_-DescriptorImpl>
``` 

I have removed no field because of the List isn't a field. I have just moved the computation from the filter to the descriptor cause I want to reload the list each time I change the configuration.

gwynjudd commented:
I've tried this fix out in our Jenkins server and it seems to resolve the issue I was seeing

escoem commented:
thanks @oleg-nenashev but IIUC, is not required because of it's populated inside the constructor:

```
          allowedOriginsList = createAllowedOriginsList(allowedOrigins);
```

Anyway, I'm going to add the readResolve method as per your request.

oleg-nenashev commented:
@escoem XStream does not invoke constructors during de-serialization of objects. So having a constructor is not enough

oleg-nenashev commented on src/main/java/org/jenkinsci/plugins/corsfilter/AccessControlsFilter.java:
```diff
@@ -182,6 +183,13 @@ void reloadAllowedOriginsList() {
         }
 
+        public Object readResolve() throws ObjectStreamException {
+            if (allowedOriginsList == null) {
```
No need to check for null, it will be always null in all correct use-cases. Better to always update it

escoem commented:
please @oleg-nenashev

dwnusbaum commented:
I tested the latest version of the PR to make sure the data was loaded correctly after a restart and it looked ok. @oleg-nenashev and @christ66 I think your review comments have been addressed.

@uaarkoti Are you still maintaining this plugin?
